### DISCUSSION TITLE
What does this constant refer to in spring_layout?

### DISCUSSION THREAD

**Discussion by anabh4569** [networkx/networkx#5328] [Q&A] [answered]
[Here](https://github.com/networkx/networkx/blob/5dfd57af2a141a013ae3753e160180b82bec9469/networkx/drawing/layout.py#L551) and in the sparse version too, there's a line that goes: `length = np.where(length < 0.01, 0.1, length)`

I have two questions about it:
1. Is this related to the enforced minimum distance of 0.01 referred to [here](https://github.com/networkx/networkx/blob/5dfd57af2a141a013ae3753e160180b82bec9469/networkx/drawing/layout.py#L543)? If so, why do we enforce this minimum distance twice?
2. Why is the constant in the first argument, `0.01`, different than the constant in the second argument `0.1`?

**Comment by dschult** [comment, marked as answer]
1. The two distances are different. The first in the code is `distance` between the nodes. The second is the length of the vectors representing displacement of those positions of the nodes. Both need to be clipped to make good layouts. They are related in that one is computed from the other. But the clipping process is not directly related which is why it is done to each separately. 
2. These values are hardcoded -- that is, not provided as arguments to the function. This is a restriction in flexibility in order to provide a reasonably understandable interface. The values work well. And that's why they are the way they are. But that being said, they were put in place in 2009 and not changed since. If you'd like to explore the impact of those values maybe a new set of eyes after 13 years can find a better answer.  It is even possible that the larger value is just a typo of 0.01, but I very much doubt it. 

I'm sorry that I don't have a better answer for you -- the short answer is that these clippings are tweaks to avoid bad layouts.
